File: sdk/ios/SsdidCore/Sources/SsdidCore/Api/DeviceApi.swift
import Foundation

/// Facade for multi-device management operations.
public struct DeviceApi {
    private let manager: DeviceManager

    init(manager: DeviceManager) {
        self.manager = manager
    }

    public func initiatePairing(identity: Identity) async throws -> PairingData {
        try await manager.initiatePairing(identity: identity)
    }

    public func joinPairing(
        did: String,
        pairingId: String,
        challenge: String,
        identity: Identity,
        deviceName: String
    ) async throws -> String {
        try await manager.joinPairing(
            did: did,
            pairingId: pairingId,
            challenge: challenge,
            identity: identity,
            deviceName: deviceName
        )
    }

    public func checkPairingStatus(did: String, pairingId: String) async throws -> PairingStatusResponse {
        try await manager.checkPairingStatus(did: did, pairingId: pairingId)
    }

    public func approvePairing(identity: Identity, pairingId: String) async throws {
        try await manager.approvePairing(identity: identity, pairingId: pairingId)
    }

    public func listDevices(identity: Identity) async throws -> [DeviceInfo] {
        try await manager.listDevices(identity: identity)
    }

    /// Revoke a device key. Currently throws ``DeviceManagerError/notSupported(_:)``
    /// because full key revocation requires multi-device key management support.
    public func revokeDevice(identity: Identity, targetKeyId: String) async throws {
        try await manager.revokeDevice(identity: identity, targetKeyId: targetKeyId)
    }
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/api/DeviceApi.kt
package my.ssdid.sdk.api

import my.ssdid.sdk.domain.device.DeviceInfo
import my.ssdid.sdk.domain.device.DeviceManager
import my.ssdid.sdk.domain.device.PairingData
import my.ssdid.sdk.domain.model.Identity
import my.ssdid.sdk.domain.transport.dto.PairingStatusResponse

class DeviceApi internal constructor(private val manager: DeviceManager) {
    suspend fun initiatePairing(identity: Identity): Result<PairingData> =
        manager.initiatePairing(identity)
    suspend fun joinPairing(
        did: String,
        pairingId: String,
        challenge: String,
        identity: Identity,
        deviceName: String
    ): Result<String> = manager.joinPairing(did, pairingId, challenge, identity, deviceName)
    suspend fun checkPairingStatus(did: String, pairingId: String): Result<PairingStatusResponse> =
        manager.checkPairingStatus(did, pairingId)
    suspend fun approvePairing(identity: Identity, pairingId: String): Result<Unit> =
        manager.approvePairing(identity, pairingId)
    suspend fun listDevices(identity: Identity): Result<List<DeviceInfo>> =
        manager.listDevices(identity)
    suspend fun revokeDevice(identity: Identity, targetKeyId: String): Result<Unit> =
        manager.revokeDevice(identity, targetKeyId)
}
